core/lib/Cardano/Multisig.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Cardano.Multisig
+    (
+      Script (..)
+    , VerificationKeyHash (..)
+    , fromVerificationKey
+    , verificationKeyHashFromBytes
+    , toScriptHash
+
+    -- * Internal
+    , toCBOR
+    ) where
+
+import Prelude
+
+import Cardano.Address
+    ( invariantSize )
+import Cardano.Address.Derivation
+    ( Depth (..), XPub )
+import Cardano.Address.Style.Shelley
+    ( Shelley, blake2b224, hashSize )
+import Control.DeepSeq
+    ( NFData )
+import Crypto.Hash
+    ( hash )
+import Crypto.Hash.Algorithms
+    ( Blake2b_224 )
+import Data.ByteString
+    ( ByteString )
+import Data.Foldable
+    ( foldl' )
+import Data.Word
+    ( Word8 )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Codec.Cbor as CBOR
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Data.ByteArray as BA
+
+
+-- | A 'Script' type represents multi signature script. The script embodies conditions
+-- that need to be satisfied to make it valid.
+--
+-- @since 3.0.0
+data Script =
+      RequireSignatureOf VerificationKeyHash
+    | RequireAllOf [Script]
+    | RequireAnyOf [Script]
+    | RequireMOf Word8 [Script]
+    deriving stock (Generic, Show, Eq)
+
+instance NFData Script
+
+-- | A 'VerificationKeyHash' type represents verification key hash
+-- that participate in building MultisigScript. The hash is expected to have size of
+-- 28-byte.
+--
+-- @since 3.0.0
+newtype VerificationKeyHash = VerificationKeyHash ByteString
+    deriving (Generic, Show, Eq)
+
+instance NFData VerificationKeyHash
+
+fromVerificationKey :: Shelley 'MultisigK XPub -> VerificationKeyHash
+fromVerificationKey = VerificationKeyHash . blake2b224
+
+verificationKeyHashFromBytes :: ByteString -> VerificationKeyHash
```

Comment:
I would either make the type failable (`Maybe VerificationKeyHash`), or call it `unsafeVerificationKeyHashFromBytes`.

Reply:
done in https://github.com/input-output-hk/cardano-addresses/pull/76/commits/b6537636e737b126cd8440a522451df694129c2e